### DISCUSSION TITLE
Assign data from drones to variables while using swarm lib and lamda function

### DISCUSSION THREAD

**Discussion by Morgens2** [bitcraze/discussions#1277] [Q&A]
Hi there!
I am working on a project using 2 crazyflie drones and the lighthouse V2 positioning system. I 've managed to print each drone's data (position, velocity etc.) through the lamda function. Now, I want to assign each drone's data to variables, in order to use them in my script. For example;   

| Variable | Corresponding data |
|--------|--------|
| X_1 | the position of the 1st crazyflie on the x axis |
| X_2 | the position of the 2nd crazyflie on the x axis |
| VX_1 | the velocity of the 1st crazyflie on the x axis |
| VX_2 | the velocity of the 2nd crazyflie on the x axis | 

Is this data already stored somewhere as a variable? And if so, how can I call/use it? 
If not, what changes does my code need in order to achive that?
Can someone please help me it figure out? Thank you in advance.

Here is the code I use for printing each drone's data.
```python
def run_shared_sequence(scf):

    box_size = 0.5
    flight_time = 2

    commander = scf.cf.high_level_commander

    commander.takeoff(0.6, 2.0)
    time.sleep(3)

    commander.go_to(box_size, 0, 0, 0, flight_time, relative=True)
    time.sleep(flight_time)

    commander.go_to(0, box_size, 0, 0, flight_time, relative=True)
    time.sleep(flight_time)

    commander.go_to(-box_size, 0, 0, 0, flight_time, relative=True)
    time.sleep(flight_time)

    commander.go_to(0, -box_size, 0, 0, flight_time, relative=True)
    time.sleep(flight_time)

    commander.land(0.0, 2.0)
    time.sleep(2)

    commander.stop()
    
def Position_callback(uri, timestamp, data, logconf):

    X = data['stateEstimate.x']
    Y = data['stateEstimate.y']
    Z = data['stateEstimate.z']

    print('{}, Position, x:  , {:.4f} ,  y:  , {:.4f} ,  z:  , {:.4f} ,'.format(uri, x, y, z), end = " " )

def PosTarget_callback(uri, timestamp, data, logconf):

    Targ_x = data['posCtl.targetX']
    Targ_y = data['posCtl.targetY']
    Targ_z = data['posCtl.targetZ']
    
    print(' Pos_Target, Cx:  , {:.4f}  , Cy:  , {:.4f}  , Cz:  ,  {:.4f} ,'.format(Targ_x, Targ_y, Targ_z), end = " ")   
    
def Velocity_callback(uri, timestamp, data, logconf):

    Vx = data['stateEstimate.vx']
    Vy = data['stateEstimate.vy']
    Vz = data['stateEstimate.vz']

    print(' Velocity, Vx:  , {:.6f} ,  Vy:  , {:.6f} ,  Vz:  , {:.6f} ,'.format(Vx, Vy, Vz), end = " " )
    
def VelTarget_callback(uri, timestamp, data, logconf):

    Targ_Vx = data['posCtl.targetVX']
    Targ_Vy = data['posCtl.targetVY']
    Targ_Vz = data['posCtl.targetVZ']

    print(' Vel_Target, CVx:  , {:.6f} ,  CVy:  , {:.6f} ,  CVz:  , {:.6f} ,'.format(Targ_Vx, Targ_Vy, Targ_Vz))
    
    
def start_position_printing(scf):
    uri = scf.cf.link_uri
    log_conf1 = LogConfig(name='Position', period_in_ms=200)
    log_conf1.add_variable('stateEstimate.x', 'float')
    log_conf1.add_variable('stateEstimate.y', 'float')
    log_conf1.add_variable('stateEstimate.z', 'float')    
    
    log_conf2 = LogConfig(name='Position_Target', period_in_ms=200)
    log_conf2.add_variable('posCtl.targetX', 'float')
    log_conf2.add_variable('posCtl.targetY', 'float')
    log_conf2.add_variable('posCtl.targetZ', 'float')
        
    log_conf3 =  LogConfig(name='Velocity', period_in_ms=200)
    log_conf3.add_variable('stateEstimate.vx', 'float')
    log_conf3.add_variable('stateEstimate.vy', 'float')
    log_conf3.add_variable('stateEstimate.vz', 'float')
    
    log_conf4 =  LogConfig(name='Velocity_Target', period_in_ms=200)
    log_conf4.add_variable('posCtl.targetVX', 'float')
    log_conf4.add_variable('posCtl.targetVY', 'float')
    log_conf4.add_variable('posCtl.targetVZ', 'float')


    scf.cf.log.add_config(log_conf1)
    scf.cf.log.add_config(log_conf2)
    scf.cf.log.add_config(log_conf3)
    scf.cf.log.add_config(log_conf4)

    
    log_conf1.data_received_cb.add_callback( lambda timestamp, data, logconf: Position_callback(scf.cf.link_uri, timestamp, data, logconf))
    log_conf2.data_received_cb.add_callback( lambda timestamp, data, logconf: PosTarget_callback(scf.cf.link_uri, timestamp, data, logconf))
    log_conf3.data_received_cb.add_callback( lambda timestamp, data, logconf: Velocity_callback(scf.cf.link_uri, timestamp, data, logconf))
    log_conf4.data_received_cb.add_callback( lambda timestamp, data, logconf: VelTarget_callback(scf.cf.link_uri, timestamp, data, logconf))

    log_conf1.start()
    log_conf2.start()
    log_conf3.start()
    log_conf4.start()
    
# |Position|-|Position_Target|-|Velocity|-|Velocity_Target|

if __name__ == '__main__':
    cflib.crtp.init_drivers()
    factory = CachedCfFactory(rw_cache='./cache')
    with Swarm(uris, factory=factory) as swarm:
        swarm.reset_estimators()
        swarm.parallel_safe(start_position_printing)
        swarm.parallel_safe(run_shared_sequence)
```

**Comment by knmcguire** [comment]
Hi! So assigning the logging variables you have done so far, but I guess you would like the variables to become available to the other drones if I'm correct?

I would advise to do the following:
-  to put the functions you made in a class, such that you can assign these variables as self.X_1, self,X_2 etc. 
- in the log callbacks, you can make an ifstatement that looks at the cfclient URI and assign the logging variables to the right script variables in the script accordingly to the drones
- Do double check it with printouts in the `run_shared_sequences` if these variables are being written properly

So these things I proposed are perhaps more python related than cflib related so for the specifics it's probably worth a google. But let me know if you don't manage to get something working and I can see if I can provide a minimal example in our cflib

**Reply by Morgens2** [reply]
Hello again! 
Firstly, thanks for answering my message. My goal is to create a function that can do math operations using each drone's data. 
So I tried to follow your advise but I got nowhere. If it's easy for you, I would love to see a minimal example. It would be very helpful for me to understand how to make the variables available for operations.
Thank you in advance.

**Reply by knmcguire** [reply]
Yeah... I think it should be easy to do (famous last words) but the problem is mostly time and priorities. None the less I've opened up an issue as it would sound like an [interesting addition](https://github.com/bitcraze/crazyflie-lib-python/issues/450), but we don't know when we will have time for it. 

As I said before, these things I've proposed are more python related problems and not cflib specific, so I do advise also doing a bit of own digging on the web, perhaps just taking my tips as guidelines as they were intended.

**Reply by Morgens2** [reply]
I totally understand. I will be waiting for the addition since it will help me with my project. Meanwhile I will do my own digging as you advised and focus on other tasks. Thanks a lot for your time and answers!

**Reply by knmcguire** [reply]
no prob! I'll update you here if we managed to get an example running, but also update us if you have done it yourself already

**Reply by Pelochus** [reply]
Hi there! I'm in a similar situation:

From a function running in parallel_safe, I want to extract the positions of all the involved Crazyflies.
I thought it could be done using `swarm.get_estimated_positions()`, but I think there is no way of calling it from inside a `parallel_safe(func)` function without modifying the library.

Any suggestions for doing this?

**Reply by knmcguire** [reply]
Hi, unfortunately no other suggestions than the one we have given above. We haven't had the time to work on [an example](https://github.com/bitcraze/crazyflie-lib-python/issues/450), so we hope that someone would like to pick that up perhaps.

**Reply by Pelochus** [reply]
Don't worry, I've already found a solution, perhaps not the best one though:

- Create a global dictionary/array for the positions you want to save
- Inside main, create a thread with a function that calls `get_estimated_positions()` every t seconds, that updates the global variable:
```python
import threading

def recover_positions():
    # Put your code here

recover_positions_thread = threading.Thread(target=recover_positions)
```
- After this, just everything else as normal, call `swarm.parallel_safe()`

**Reply by knmcguire** [reply]
Just mind that the get_estimated_position() is reported to give issues if it is used in a loop for too long. That is because it practically starts up a new logblock every time it gets called, which will create issues in the long run (https://github.com/bitcraze/crazyflie-lib-python/blob/c7eaabe31612e796fcdcf8511c18d197051e17cd/cflib/crazyflie/swarm.py#L115)

It might be better to setup async logging yourself based on the examples.

**Reply by Pelochus** [reply]
Thanks, found that I had some problems, perhaps it is related to this.

By the way, wouldn't the logblock be destroyed at the end of the function? Not very familiar with Python, but I would expect that from other languages.

**Reply by knmcguire** [reply]
No that is the problem, there is no functionality to destroy the logblock, as it will persist within the crazyflie. 

I advise you to take a look at this logging example. You can access each of the crazyflie instances in the swarm class so you can set this up per crazyflie as well: https://github.com/bitcraze/crazyflie-lib-python/blob/master/examples/logging/basiclog.py
